File: crates/yokan/demo/lookup.py
# /// script
# requires-python = ">=3.14"
# ///
"""Dict cells. The order question is DECIDED: iteration stays out
(Python orders by insertion, native maps by key — admitting either
would lie), and everything order-free is in: per-key writes land in
place in both tiers (`prices["cherry"] = 200` is pixie's
`m[k] = v`), reads are total via .get(key, default), membership
guards conditions, len counts.
"""
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

from yokan import button, column, row, run, State, text  # noqa: E402

prices: State[dict[str, int]] = State({"apple": 120, "banana": 80})
picked: State[int] = State(0)
label: State[str] = State("none")


def pick_apple():
    picked.set(prices().get("apple", -1))
    if "cherry" in prices():
        label.set("cherry known")
    else:
        label.set("no cherry")


def add_cherry():
    prices["cherry"] = 200
    picked.set(prices().get("cherry", -1))
    if "cherry" in prices():
        label.set("cherry known")


def miss():
    picked.set(prices().get("durian", -7))


def view():
    with column(spacing=8, padding=12):
        text(f"picked={picked()} n={len(prices())} {label()}")
        text(f"apple costs {prices().get('apple', -1)} right now", size=12)
        with row(spacing=6):
            button("apple", on_click=pick_apple)
            button("cherry", on_click=add_cherry)
            button("miss", on_click=miss)


if __name__ == "__main__":
    run(view, title="lookup")

File: gomamochi/demo/lookup.go
// A map on the app: reading with a fallback, asking whether a key is
// there, and adding one while the window is open. A map is never
// ranged over in a view — its order changes from run to run — but a
// key is read as freely as a field.
package main

import (
	"fmt"

	. "github.com/i2y/yokan/gomamochi"
)

type Lookup struct {
	prices map[string]int
	picked int
	label  string
}

// The price of a fruit, or `fallback` when it is not on the list.
func (l *Lookup) price(name string, fallback int) int {
	if v, ok := l.prices[name]; ok {
		return v
	}
	return fallback
}

func (l *Lookup) pickApple() {
	l.picked = l.price("apple", -1)
	if _, ok := l.prices["cherry"]; ok {
		l.label = "cherry known"
	} else {
		l.label = "no cherry"
	}
}

func (l *Lookup) addCherry() {
	l.prices["cherry"] = 200
	l.picked = l.price("cherry", -1)
	if _, ok := l.prices["cherry"]; ok {
		l.label = "cherry known"
	}
}

func (l *Lookup) View() Element {
	return Column(
		Text(fmt.Sprintf("picked=%d n=%d %s", l.picked, len(l.prices), l.label)),
		Text(fmt.Sprintf("apple costs %d right now", l.price("apple", -1))).Size(12),
		Row(
			Button("apple").OnClick(func() { l.pickApple() }),
			Button("cherry").OnClick(func() { l.addCherry() }),
			Button("miss").OnClick(func() { l.picked = l.price("durian", -7) }),
		).Spacing(6),
	).Spacing(8).Padding(12)
}

func main() {
	Run(&Lookup{prices: map[string]int{"apple": 120, "banana": 80}, label: "none"}, Title("lookup"))
}
